components/suggest/src/db.rs

```diff
@@ -197,58 +183,119 @@ impl<'a> SuggestDao<'a> {
+                ":provider": SuggestionProvider::Wikipedia
+            },
+            |row| -> Result<Suggestion> {
+                let suggestion_id: i64 = row.get("id")?;
+                let title = row.get("title")?;
+                let raw_url = row.get::<_, String>("url")?;
+
+                let keywords: Vec<String> = self.conn.query_rows_and_then_cached(
+                    "SELECT keyword FROM keywords
+                     WHERE suggestion_id = :suggestion_id AND rank >= :rank
+                     ORDER BY rank ASC",
+                    named_params! {
+                        ":suggestion_id": suggestion_id,
+                        ":rank": row.get::<_, i64>("rank")?,
+                    },
+                    |row| row.get(0),
+                )?;
+                let icon = self.conn.try_query_one(
+                    "SELECT i.data
+                     FROM icons i
+                     JOIN wikipedia_custom_details s ON s.icon_id = i.id
+                     WHERE s.suggestion_id = :suggestion_id",
+                    named_params! {
+                        ":suggestion_id": suggestion_id
+                    },
+                    true,
+                )?;
+                Ok(Suggestion::Wikipedia {
+                    title,
+                    url: raw_url,
+                    full_keyword: full_keyword(keyword_lowercased, &keywords),
+                    icon,
+                })
+            },
+        )?;
+        Ok(suggestions)
+    }
+
+    /// Fetches Suggestions of type Amo provider that match the given query
+    pub fn fetch_amo_suggestions(
+        &self,
+        query: &SuggestionQuery,
+        suggestions_limit: i32,
+    ) -> Result<Vec<Suggestion>> {
+        let keyword_lowercased = &query.keyword.to_lowercase();
+        let (keyword_prefix, keyword_suffix) = split_keyword(keyword_lowercased);
+        let suggestions = self.conn.query_rows_and_then_cached(
+            "SELECT s.id, k.rank, s.title, s.url, s.provider, s.score, k.confidence, k.keyword_suffix
+                     FROM suggestions s
+                     JOIN prefix_keywords k ON k.suggestion_id = s.id
+                     WHERE k.keyword_prefix = :keyword_prefix AND s.provider = :provider
+                     ORDER by s.score DESC
+                     LIMIT :suggestions_limit",
+             named_params! {
+                ":keyword_prefix": keyword_prefix,
+                ":provider": SuggestionProvider::Amo,
+                ":suggestions_limit": suggestions_limit,
+            },
+            |row| -> Result<Option<Suggestion>>{
+                let suggestion_id: i64 = row.get("id")?;
+                let title = row.get("title")?;
+                let raw_url = row.get::<_, String>("url")?;
+                let score = row.get::<_, f64>("score")?;
+
+                let full_suffix = row.get::<_, String>("keyword_suffix")?;
+                    if !full_suffix.starts_with(keyword_suffix){
```

Comment:
It looks like misaligned also there is a missing whitespace `suffix) {`. I opted for `bool::then()` [here](https://mozilla.github.io/application-services/book/howtos/branch-builds.html#breaking-changes-in-an-application-services-branch) to cut an if-else nesting, it's up to you though :)